src/representations/representation_manager_spherical_covariants.hh

```diff
@@ -0,0 +1,500 @@
+    std::array<std::shared_ptr<internal::SphericalInvariantsPrecomputationBase>,
+               internal::enumSize<internal::SphericalCovariantsType>()>
+        precompute_spherical_covariants{};
+    std::string spherical_covariants_type_str{};
+    std::vector<Precision_t> dummy{};
+    bool inversion_symmetry{false};
+    size_t lambda{0};
+    bool normalize{true};
+  };
+
+  template <class Mngr>
+  void RepresentationManagerSphericalCovariants<Mngr>::compute() {
+    using internal::SphericalCovariantsType;
+    switch (this->spherical_covariants_type) {
+    case SphericalCovariantsType::LambdaSpectrum:
+      this->compute_lambdaspectrum();
+      break;
+    default:
+      // Will never reach here (it's an enum...)
+      break;
+    }
+  }
+
+  template <class Mngr>
+  void RepresentationManagerSphericalCovariants<
+      Mngr>::initialize_percenter_lambda_soap_vectors() {
+    using math::pow;
+
+    size_t n_row{pow(this->max_radial, 2_z)};
+    // number of combinations of l1 and l2 satisfying the triangle constraint
+    size_t n_col{0};
+    if (this->inversion_symmetry == false) {
+      n_col = static_cast<size_t>(
+          (2 + this->lambda - 3 * pow(this->lambda, 2_z) +
+           2 * this->max_angular + 4 * this->lambda * this->max_angular) /
+          2 * (2 * this->lambda + 1));
+    } else {
+      n_col =
+          static_cast<size_t>(
+              std::ceil(pow(this->max_angular + 1, 2_z) / 2.0) -
+              pow(1.0 + std::floor((this->lambda - 1) / 2.0), 2) -
+              std::floor(pow(this->max_angular + 1 - this->lambda, 2_z) / 2.0) *
+                  (this->lambda % 2) -
+              (std::ceil(pow(this->max_angular + 1 - this->lambda, 2_z) / 2.0) -
+               (this->max_angular - this->lambda + 1)) *
+                  (1.0 - this->lambda % 2)) *
+          (2 * this->lambda + 1);
+      if (this->lambda % 2 == 1) {
+        n_col = static_cast<size_t>(
+            0.5 *
+                (2 + this->lambda - 3 * pow(this->lambda, 2_z) +
+                 2 * this->max_angular + 4 * this->lambda * this->max_angular) *
+                (2 * this->lambda + 1) -
+            static_cast<int>(n_col));
+      }
+    }
+
+    // clear the data container and resize it
+    this->soap_vectors.clear();
+    this->soap_vectors.set_shape(n_row, n_col);
+    this->soap_vectors.resize();
+
+    auto & expansions_coefficients{rep_expansion.expansions_coefficients};
+
+    // identify the species in each environment and initialize soap_vectors
+    for (auto center : this->structure_manager) {
+      auto & coefficients{expansions_coefficients[center]};
+      auto & soap_vector{this->soap_vectors[center]};
+      internal::Sorted<false> is_not_sorted{};
+
+      std::vector<internal::SortedKey<Key_t>> pair_list{};
+      auto & center_type{center.get_atom_type()};
+      Key_t pair_type{center_type, center_type};
+      // TODO(felix) optimize this loop
```

Comment:
the species are not sorted by construction so there are sorted explicitly and many redundant combinations are present in pair_list. This is not a super substantial speed up since it increases the number map query to see whether a key has already been registered but it could be optimized.